File: javascript/3606.Coupon-Code-Validator.js
/**
 * @param {string[]} code
 * @param {string[]} businessLine
 * @param {boolean[]} isActive
 * @return {string[]}
 */
var validateCoupons = function (code, businessLine, isActive) {
  const order = {
    electronics: 0,
    grocery: 1,
    pharmacy: 2,
    restaurant: 3,
  };

  const validCodeRegex = /^[A-Za-z0-9_]+$/;

  const validCoupons = [];

  for (let i = 0; i < code.length; i++) {
    if (
      validCodeRegex.test(code[i]) &&
      order.hasOwnProperty(businessLine[i]) &&
      isActive[i]
    ) {
      validCoupons.push({
        code: code[i],
        businessLine: businessLine[i],
      });
    }
  }

  validCoupons.sort((a, b) => {
    const lineDiff = order[a.businessLine] - order[b.businessLine];
    if (lineDiff !== 0) {
      return lineDiff;
    }

    if (a.code < b.code) return -1;
    if (a.code > b.code) return 1;
    return 0;
  });

  return validCoupons.map((coupon) => coupon.code);
};

File: python/3606.Coupon-Code-Validator.py
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        order = {
            "electronics": 0,
            "grocery": 1,
            "pharmacy": 2,
            "restaurant": 3,
        }

        valid_code_regex = re.compile(r'^[A-Za-z0-9_]+$')

        valid_coupons = []

        for i in range(len(code)):
            if valid_code_regex.match(code[i]) and businessLine[i] in order and isActive[i]:
                valid_coupons.append({
                    "code": code[i],
                    "businessLine": businessLine[i]
                })

        valid_coupons.sort(key=lambda x: (order[x["businessLine"]], x["code"]))

        return [coupon["code"] for coupon in valid_coupons]